### PR TITLE
MAINT: Change the type of error raised in set_printoptions

### PR DESCRIPTION
**Pull request by kianasun**

### CODE REVIEW

**Comment by eric-wieser** [comment]
I don't think this is an improvement - passing `np.nan` should not give a `TypeError` if passing `1.0` does not, because they have the same type.

If you want to go down this path, then you should split the condition in two, and raise `TypeError` for non-numeric types, and `ValueError` for nans

**Comment by kianasun** [comment]
@eric-wieser You are right. I have split it into two conditions.

**Review by eric-wieser — APPROVED** [review]
@mattip, think the message is still clear enough for clueless stackoverflow users using `threshold='nan'`?

**Comment by charris** [comment]
Needs a release note because of changed behavior.

**Comment by kianasun** [comment]
Hi @charris , I added a release note.

**Comment by kianasun** [comment]
@charris @eric-wieser @mattip Anything else need to be changed?

**Comment by eric-wieser** [comment]
I think it would be best to wait for the #13908 stuff to go in first - your release note is in the wrong place, but the right place is about to change.

**Comment by seberg** [comment]
@kianasun can you move your release note into the file `changelog/13899.changes.rst`, it will be a very short file containing only what you wrote. I think after that happens we can probably just merge it, so just ping me on that (I can do that if you like, but would be happy if you can move it).

**Comment by kianasun** [comment]
@seberg Sure! will do this soon.

File: changelog/13899.changes.rst
Comment (seberg) [inline changelog/13899.changes.rst]:
```suggestion
Previously an incorrect ``threshold`` raised ``ValueError``; it now raises ``TypeError``
```
Actually, now I am confused about the right grammar. Release notes get a final pass in any case though, so will merge without the fixup in a bit either way.

Comment (kianasun) [reply changelog/13899.changes.rst]: you are right, it should be `raised`

**Comment by seberg** [comment]
Thanks @kianasun!

**Pull request merged**
